internal/controller/management/upgrade/credentials.go

```diff
@@ -0,0 +1,123 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	SecretTypeLabelKey  = "kargo.akuity.io/secret-type" // nolint: gosec
+	repoLabelValue      = "repository"
+	repoCredsLabelValue = "repo-creds" // nolint: gosec
+)
+
+// credentialsReconciler reconciles credentials (Secrets) to upgrade them from
+// v0.4.0-compatible to v0.5.0-compatible.
+type credentialsReconciler struct {
```

Comment:
My worry with updating the Secrets for the user is that you end up fighting a GitOps agent.

Given this, I am wondering if supporting both formats while emitting some sort of warning while the "legacy" format is encountered wouldn't be the better approach.

Reply:
> My worry with updating the Secrets for the user is that you end up fighting a GitOps agent.

That's a really good point.

> Given this, I am wondering if supporting both formats while emitting some sort of warning while the "legacy" format is encountered wouldn't be the better approach.

The logical place to surface this would be in a Warehouse's `status.Message`. If things are working, I don't think people will ever look there. They'll never see the warning.

What would you think about removing this conversion due to the gitops issue and letting things _fail_ when the old format is encountered? That seems more likely to prompt someone to look at the Warehouse's `status.Message` field.

This means we need to at least _look_ for older credentials, but don't need to use them.

Reply:
> My worry with updating the Secrets for the user is that you end up fighting a GitOps agent.

Alternatively, that may not be the end of the world. People will notice that and fix it.

Reply:
@jessesuen do you have any opinions about upgrading credential formats?

Reply:
I would think people look at their logs at least once in a while in combination with reading the release notes, making this the least invasive and user friendly approach in my opinion.

But I can live with any option you think is good. Just wanted to highlight it as an alternative.

Reply:
> in combination with reading the release notes

☝️ Leveraging those might be my own preference.

Instead of "Hey, we broke all your creds." It's, "Hey, we broke all your creds and then fixed them. If you're gitops'ing your projects, you might see some odd behavior, but you know how to fix it."

Overall, since we're not GA, I don't want to _overthink_ the upgrade path from v0.4.x to v0.5.x. I'm happy as long as we don't leave anyone in a degraded state that they cannot bound back from and I think we've at least ticked those boxes.

Reply:
> @jessesuen do you have any opinions about upgrading credential formats?

I don't have as much of a strong opinion here, since credentials are an easier and much more obvious thing to manually fix, than how we broke freights. Migration of credentials would be nice to have, IMO but not as required as the other stuff we broke.

Reply:
I mean... it's done already. It's purely a matter of whether you think we ought _not_ touch secrets, which is a perspective I could understand.